quantuminspire/settings/base_settings.py

```diff
@@ -0,0 +1,184 @@
+        raise NotImplementedError
+
+    def __call__(self) -> Any:
+        encoding = self.config.get("env_file_encoding")
+
+        ensure_config_file_exists(self._json_file_path, encoding, self._default_factory)
+        return json.loads(self._json_file_path.read_text(encoding))
+
+
+class BaseConfigSettings(BaseSettings, ABC):  # pylint: disable=too-few-public-methods
+    """The settings class for the Quantum Inspire persistent configuration."""
+
+    model_config = SettingsConfigDict(
+        env_file_encoding="utf-8",
+        env_prefix="QI2_",
+    )
+
+    backend_type: Optional[int] = Field(None)
+
+    @classmethod
+    def json_file(cls) -> Path:
+        """Return the full JSON file path (base_dir + subpath)."""
+        return Path.joinpath(cls.base_dir(), ".quantuminspire", "config.json")
+
+    @classmethod
+    @abstractmethod
+    def base_dir(cls) -> Path:
+        """Subclasses provide the base directory (home, cwd, etc)"""
+        raise NotImplementedError
+
+    @classmethod
+    @abstractmethod
+    def default_factory(cls) -> Dict[str, Any]:
+        """Return default JSON content."""
+        raise NotImplementedError
+
+    # R0913: Too many arguments (6/5) (too-many-arguments)
+    @classmethod
+    def settings_customise_sources(  # pylint: disable=R0913
+        cls,
+        settings_cls: Type[BaseSettings],
+        init_settings: PydanticBaseSettingsSource,
+        env_settings: PydanticBaseSettingsSource,
+        dotenv_settings: PydanticBaseSettingsSource,
+        file_secret_settings: PydanticBaseSettingsSource,
+    ) -> Tuple[PydanticBaseSettingsSource, ...]:
+        """Customise the settings sources (by adding, removing and changing the order of sources).
+
+        Args:
+            init_settings: The initial settings (Settings object creation): highest priority.
+            env_settings: The configuration settings (Config inner object creation).
+            file_secret_settings: The file secret settings: lowest priority
+
+        Returns:
+            The original sources, with
+            - the JSON file as source added after the env settings and before the file secret settings.
+            The order determines the priority!
+        """
+        return (
+            init_settings,
+            env_settings,
+            JsonConfigSettingsSource(settings_cls),
+            file_secret_settings,
+        )
+
+    def _get_parent_and_attr(self, key: str) -> Tuple[BaseModel, str]:
+        """Return the parent object and final attribute name for a dot-separated key."""
+
+        parts = key.split(".")
+        current: BaseModel = self
+
+        for part in parts[:-1]:
+            current = getattr(current, part)
+        return current, parts[-1]
+
+    def clear(self, key: str) -> None:
+        """Clear a setting value.
+
+        Supports nested keys using dot notation, e.g., 'project.description'.
```

Comment:
This example is a key, right? It would be something like: "project.description = Trying out a new implementation of the BB8 algorithm"? It's not that `project` is a placeholder for the directory name?